File: reverse/upos/upos.py
import hashlib
import itertools
import subprocess
import base64
import math

class Streamer:
    def __init__(self, seed="01101000010", tap=8):
        self.lfsr = [False if bit == '0' else True for bit in seed]
        self.tap = (len(seed) - 1) - tap

    def step(self):
        new_bit = self.lfsr[self.tap] ^ self.lfsr[0]
        for i in range(len(self.lfsr) - 1):
            self.lfsr[i] = self.lfsr[i + 1]
        self.lfsr[-1] = new_bit
        return 0 if not new_bit else 1

    def g2(self):
        val = 0
        for i in range(16):
            val |= self.step() << i
        return val

    def generate(self, k):
        temp = 0
        for i in range(k):
            temp = (temp * 2) + self.step()
        return temp

    def __str__(self):
        representation = ''.join(['1' if bit else '0' for bit in self.lfsr])
        return representation


class Upos:
    def __init__(self):
        init_vector = [byte + 128 for byte in [-34, -83, -66, -17, -34, -83, -66, -17, -34, -83, -66, -17, -34, -83, -66, -17]]
        self.init_vector = bytearray(init_vector)
        # print(sum_array)

    def main(self):

        characters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_!?-.|~`@#$%^&*(+=-)'

        
        m = [[0] * 256 for _ in range(256)]
        instance = Upos()
        instance.lm(m)

        
        idx16 = 100

        ip_bool = [Upos.ip(i2) * i2 for i2 in range(30)]

        print(f"ip_bool {ip_bool}")

        sum_array = [sum(ip_bool[:i+1]) + 12 for i in range(len(ip_bool))]
        for i in range(30):
            sum_array[i] = sum_array[i] + 32 * i
        print(f"sum_array {sum_array}")



        # s = Streamer()
        # for _ in range(12):
        #     s.step()
        flag = ""


        for i2 in range(30):
            print(f"round {i2}")

            for combination in itertools.product(characters, repeat=2):
                curr = ''.join(combination)

            # curr = core[i2 * 2:i2 * 2 + 2]
            
                # if Upos.ip(i2):
                #     for j in range(i2):
                #         s.step()
                s = self.streamer_step(i2, sum_array)
                j2 = s.g2()
                # print(f"j2 {j2}")
                mX = j2 & 255
                # print(f"mX {mX}")
                mY = (s.g2() & 0xFF00) >> 8
                # print(f"mY {mY}")
                # print(f"correct String {Upos.sq(Upos.r(curr))} {m[mX][mY]}")
                if Upos.sq(Upos.r(curr)) == m[mX][mY]:
                    print(f"correct String {i2} {curr}")
                    flag = flag + curr
                else:
                    idx16 += 1
        
        print(f"flag {flag}")

        # for combination in itertools.product(characters, repeat=2):
        #     curr = ''.join(combination)

        #     flag = 'MOBISEC{Isnt_this_a_truly_evil_undebuggable_piece_of_s' + curr + 'W_software??}'

        #     if self.h(flag) == "4193d9b72a5c4805e9a5cc739f8a8fc23b2890e13b83bb887d96f86c30654a12":
        #         print(f"flag {flag}")
        #         return
            


# MOBISEC{Isnt_this_a_truly_evil_undebuggable_piece_of_sh^W_software??}
           
    def streamer_step(self, i, sum_array):
        s = Streamer()
        round_number = sum_array[i]
        for _ in range(round_number):
            s.step()
        
        return s



    @staticmethod
    def ec(cmd):
        out = ""
        try:
            process = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
            stdout, _ = process.communicate()
            out = out + stdout.decode("utf-8")
        except Exception as e:
            pass
        return out

    @staticmethod
    def r(s):
        out = ""
        for c in s:
            if 'a' <= c <= 's' or 'A' <= c <= 'S':
                c = chr(ord(c) + 7)
            elif 't' <= c <= 'z' or 'T' <= c <= 'Z':
                c = chr(ord(c) - 19)
            out = out + c
        return out

    @staticmethod
    def ip(x):
        for i in range(2, x):
            if x % i == 0:
                return False
        return True

    @staticmethod
    def sq(a):

        n = (ord(a[0]) + (ord(a[1]) << 8)) & 0xFFFF
        n2 = pow(n, 2)
        return n2


    def h(self, flag):
        try:
            md = hashlib.sha256()
            md.update(flag.encode())
            digest = md.digest()
            return self.th(digest)
        except Exception as e:
            return None

    def th(bytes):
        hex_string = ''.join(format(b, '02x') for b in bytes)
        return hex_string


    def lm(self, matrix):
        try:
            with open("lotto.dat", "r") as file:
                rowIdx = 0
                for line in file:
                    elems = line.split()
                    colIdx = 0
                    for elem in elems:
                        e = int(elem)
                        matrix[rowIdx][colIdx] = e
                        colIdx += 1
                    if colIdx != 256:
                        raise Exception("Error")
                    rowIdx += 1
                if rowIdx != 256:
                    raise Exception("Error")
        except Exception as e:
            # Handle exception (e.g., print an error message)
            raise e


    @staticmethod
    def dec(x):
        return base64.b64decode(x).decode("utf-8")


if __name__ == "__main__":
    upos_instance = Upos()
    upos_instance.main()

File: reverse/upos/upos.java

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.lang.reflect.Array;
import java.security.MessageDigest;
import java.util.Base64;

import ups.streamer;


// .\jadx-gui-1.4.7.exe --comments-level debug 
public class upos {
    private static byte[] initVector = {-34, -83, -66, -17, -34, -83, -66, -17, -34, -83, -66, -17, -34, -83, -66, -17};

    // MOBISEC{this_is_           really_                              upos}
    public static void main(String[] args) throws Exception {
        // streamer lfsr = new streamer("01101000010", 8);
        // System.out.println("Testing step()");
        // for (int i = 0; i < 10; i++) {
        //     int bit = lfsr.step();
        //     PrintStream printStream = System.out;
        //     printStream.println(lfsr + " " + bit);
        // }
        // streamer lfsr2 = new streamer("01101000010", 8);
        // System.out.println("\nTesting generate()");
        // for (int i2 = 0; i2 < 10; i2++) {
        //     int r = lfsr2.generate(5);
        //     PrintStream printStream2 = System.out;
        //     printStream2.println(lfsr2 + " " + r);
        // }

        long[][] m = (long[][]) Array.newInstance(long.class, 256, 256);

        // String flag = "MOBISEC{this_is_";
        // String flagEnd = "upos}";

        // // L49f
        // String hash = "4193d9b72a5c4805e9a5cc739f8a8fc23b2890e13b83bb887d96f86c30654a12";

        upos instance = new upos();
        instance.lm(m);

        // System.out.println(upos.ip(0));

        // String r12 = "";


        // String encryptedString = "bHM=";
        // String decryptedString = instance.dec(encryptedString);
        // r12 = decryptedString + " ";

        // System.out.println("Encrypted String: " + encryptedString);
        // System.out.println("Decrypted String: " + decryptedString);

        // encryptedString = "L3Byb2M=";
        // decryptedString = instance.dec(encryptedString);

        // System.out.println("Encrypted String: " + encryptedString);
        // System.out.println("Decrypted String: " + decryptedString);

        // r12 = r12 + decryptedString;

        // String r8 = instance.ec(r12);
        // System.out.println("String: " + r8);

        // encryptedString = "Y2F0";
        // decryptedString = instance.dec(encryptedString);

        // System.out.println("Encrypted String: " + encryptedString);
        // System.out.println("Decrypted String: " + decryptedString);

        // encryptedString = "ZnJpZGE=";
        // decryptedString = instance.dec(encryptedString);

        // System.out.println("Encrypted String: " + encryptedString);
        // System.out.println("Decrypted String: " + decryptedString);

        // encryptedString = "Y29tLmFuZHJvaWQudmVuZGluZw==";
        // decryptedString = instance.dec(encryptedString);

        // System.out.println("Encrypted String: " + encryptedString);
        // System.out.println("Decrypted String: " + decryptedString);

        // encryptedString = "YW5kcm9pZC5vcy5EZWJ1Zw==";
        // decryptedString = instance.dec(encryptedString);

        // System.out.println("Encrypted String: " + encryptedString);
        // System.out.println("Decrypted String: " + decryptedString);

        // encryptedString = "aXNEZWJ1Z2dlckNvbm5lY3RlZA==";
        // decryptedString = instance.dec(encryptedString);

        // System.out.println("Encrypted String: " + encryptedString);
        // System.out.println("Decrypted String: " + decryptedString);

        // String out = instance.r("018a94a01edcfd1c8121f56dd36a412e62b3dd8");
        // System.out.println("out String: " + out);

        streamer s = new streamer();

        String core = "this_is";



        int idx16 = 100; //fs[100] - fs[129] true
        for (int i2 = 0; i2 < 1; i2++) {
            
            String curr = Character.toString(core.charAt(i2 * 2)) + Character.toString(core.charAt((i2 * 2) + 1));
            if (ip(i2)) { // starts with 14
                for (int j = 0; j < i2; j++) {
                    s.step();
                }
            }
            int j2 = s.g2();
            int mX = j2 & 255;
            int mY = (s.g2() & 65280) >> 8;
            if (sq(r(curr)) == m[mX][mY]) {
                System.out.println("correct String" + i2 + curr);
            } else {
                idx16++;
            }
        }

    }




    public static String ec(String cmd) {
        String out = new String();
        try {
            Process p = Runtime.getRuntime().exec(cmd);
            InputStream stdout = p.getInputStream();
            byte[] buffer = new byte[102400];
            while (true) {
                int read = stdout.read(buffer);
                if (read > 0 && read <= 102400) {
                    String line = new String(buffer, 0, read);
                    out = out + line;
                } else if (stdout.available() <= 0) {
                    break;
                }
            }
            stdout.close();
        } catch (Exception e) {
        }
        return out;
    }

    
    public static String r(String s) {
        String out = "";
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c >= 'a' && c <= 's') {
                c = (char) (c + 7);
            } else if (c >= 'A' && c <= 'S') {
                c = (char) (c + 7);
            } else if (c >= 't' && c <= 'z') {
                c = (char) (c - 19);
            } else if (c >= 'T' && c <= 'Z') {
                c = (char) (c - 19);
            }
            out = out + c;
        }
        return out;
    }

    public static boolean ip(int x) {
        for (int i = 2; i < x; i++) {
            if (x % i == 0) {
                return false;
            }
        }
        return true;
    }


    public static long sq(String a) {
        int n = (a.charAt(0) + (a.charAt(1) << '\b')) & 65535;
        long n2 = (long) Math.pow(n, 2.0d);
        return n2;
    }

    public static String th(byte[] bytes) { // to Hex
        StringBuilder hexString = new StringBuilder();
        for (byte b : bytes) {
            String hex = Integer.toHexString(b & 255);
            if (hex.length() == 1) {
                hexString.append('0');
            }
            hexString.append(hex);
        }
        return hexString.toString();
    }

    private void lm(long[][] matrix) throws Exception { //load me
        BufferedReader reader = null;
        try {
            BufferedReader reader2 = new BufferedReader(new InputStreamReader(new FileInputStream("lotto.dat")));
            int rowIdx = 0;
            while (true) {
                String row = reader2.readLine();
                if (row != null) {
                    String[] elems = row.split(" ");
                    int colIdx = 0;
                    for (String elem : elems) {
                        long e = Long.parseLong(elem);
                        matrix[rowIdx][colIdx] = e;
                        colIdx++;
                    }
                    if (colIdx != 256) {
                        throw new Exception("error");
                    }
                    rowIdx++;
                } else if (rowIdx != 256) {
                    throw new Exception("error");
                } else {
                    reader2.close();
                    return;
                }
            }
        } catch (Throwable th) {
            if (0 != 0) {
                reader.close();
            }
            throw th;
        }
    }

    public static String dec(String x) {
        return new String(Base64.getDecoder().decode(x));
    }
}

